src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,547 @@
+// Copyright 2018 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tls
+
+// Delegated credentials for TLS
+// (https://tools.ietf.org/html/draft-ietf-tls-subcerts) is an IETF Internet
+// draft and proposed TLS extension. If the client supports this extension, then
+// the server may use a "delegated credential" as the signing key in the
+// handshake. A delegated credential is a short lived public/secret key pair
+// delegated to the server by an entity trusted by the client. This allows a
+// middlebox to terminate a TLS connection on behalf of the entity; for example,
+// this can be used to delegate TLS termination to a reverse proxy. Credentials
+// can't be revoked; in order to mitigate risk in case the middlebox is
+// compromised, the credential is only valid for a short time (days, hours, or
+// even minutes).
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+const (
+	dcMaxTTLSeconds   = 60 * 60 * 24 * 7 // The maxium validity period is 7 days
+	dcMaxTTL          = time.Duration(dcMaxTTLSeconds * time.Second)
+	dcMaxPubLen       = 1 << 24 // Bytes
+	dcMaxSignatureLen = 1 << 16 // Bytes
+)
+
+var errNoDelegationUsage = errors.New("tls: certificate not authorized for delegation")
+var extensionDelegatedCredential = []int{1, 3, 6, 1, 4, 1, 44363, 44}
+
+// isValidForDelegation returns true if a certificate can be used for delegated
+// credentials.
+func isValidForDelegation(cert *x509.Certificate) bool {
+	// Check that the digitalSignature key usage is set.
+	// The certificate must contains the digitalSignature KeyUsage.
+	if (cert.KeyUsage & x509.KeyUsageDigitalSignature) == 0 {
+		return false
+	}
+
+	// Check that the certificate has the DelegationUsage extension and that
+	// it's non-critical (See Section 4.2 of RFC5280).
+	for _, extension := range cert.Extensions {
+		if extension.Id.Equal(extensionDelegatedCredential) {
+			return true
+		}
+
+		if !extension.Critical {
+			return true
+		}
+	}
```

Comment:
Does this achieve the intended functionality? seems you also return true for the first non-critical extension?

```suggestion
	for _, extension := range cert.Extensions {
		if extension.Id.Equal(extensionDelegatedCredential) {
			return !extension.Critical
		}
	}
```

(maybe we need to abort with an error, because the certificate is not valid at all?)

Reply:
Mmm.. well, the function where this function is called actually sends the error:

```
var errNoDelegationUsage = errors.New("tls: certificate not authorized for delegation")
```

I fix the extension check ;)